shared/src/main/scala/pl/metastack/metarouter/Router.scala

```diff
@@ -69,16 +65,17 @@ object Router {
           }
       }
 
-    m[ROUTE](route.pathElements, split(s)).map { x =>
-      // Using asInstanceOf as a band aid since compiler isn't able to confirm the type.
-      RouteData(route, x.asInstanceOf[L])
-    }
+    m[ROUTE](route.pathElements, split(uri))
   }
 
-  def parse[ROUTE <: HList, T, Args <: HList](mappedRoute: MappedRoute[ROUTE, T], uri: String)
-                                             (implicit gen: Generic.Aux[T, Args],
-                                                       map: FlatMapper.Aux[Args.Convert.type, ROUTE, Args]): Option[T] =
-    parse(mappedRoute.route, uri).map(parsed => gen.from(parsed.data))
+  def parse[ROUTE <: HList, T, Args <: HList]
+    (mappedRoute: MappedRoute[ROUTE, T], uri: String)
+    (implicit gen: Generic.Aux[T, Args],
+              map: FlatMapper.Aux[Args.Convert.type, ROUTE, Args]): Option[T] =
+    parse(mappedRoute.route, uri).map { x =>
+      // Using asInstanceOf as a band aid since compiler isn't able to confirm the type.
+      gen.from(x.asInstanceOf[Args])
```

Comment:
`route.zip(parsedUrl).filter(_._1.isInstanceOf[Arg]).map(parseChunk)` - might be a bit tricky for hlists, will require some magic with typeclasses.